### PR TITLE
Add symbol visibility information to all the functions that are exported

### PR DESCRIPTION
**Pull request by imciner2**
This adds the symbol visibility information to the main API functions that QDLDL exports so that they are exposed inside shared libraries on all platforms. This is mainly an issue on Windows (which defaults to exporting no symbols in a shared library), but could also be a problem if someone tries to use the flag `-fvisibility=hidden` on other platforms.

Before this change (compiled with `-fvisibility=hidden`), `readelf -Ws out/libqdldl.so` shows:
```
Symbol table '.dynsym' contains 5 entries:
   Num:    Value          Size Type    Bind   Vis      Ndx Name
     0: 0000000000000000     0 NOTYPE  LOCAL  DEFAULT  UND 
     1: 0000000000000000     0 NOTYPE  WEAK   DEFAULT  UND _ITM_deregisterTMCloneTable
     2: 0000000000000000     0 NOTYPE  WEAK   DEFAULT  UND __gmon_start__
     3: 0000000000000000     0 NOTYPE  WEAK   DEFAULT  UND _ITM_registerTMCloneTable
     4: 0000000000000000     0 FUNC    WEAK   DEFAULT  UND __cxa_finalize@GLIBC_2.2.5 (2)
```

After this change (compiled with `-fvisibility=hidden` still), `readelf -Ws out/libqdldl.so` shows:
```
Symbol table '.dynsym' contains 10 entries:
   Num:    Value          Size Type    Bind   Vis      Ndx Name
     0: 0000000000000000     0 NOTYPE  LOCAL  DEFAULT  UND 
     1: 0000000000000000     0 NOTYPE  WEAK   DEFAULT  UND _ITM_deregisterTMCloneTable
     2: 0000000000000000     0 NOTYPE  WEAK   DEFAULT  UND __gmon_start__
     3: 0000000000000000     0 NOTYPE  WEAK   DEFAULT  UND _ITM_registerTMCloneTable
     4: 0000000000000000     0 FUNC    WEAK   DEFAULT  UND __cxa_finalize@GLIBC_2.2.5 (2)
     5: 0000000000001240  1120 FUNC    GLOBAL DEFAULT   11 QDLDL_factor
     6: 00000000000016a0    96 FUNC    GLOBAL DEFAULT   11 QDLDL_Lsolve
     7: 0000000000001700    89 FUNC    GLOBAL DEFAULT   11 QDLDL_Ltsolve
     8: 0000000000001760   211 FUNC    GLOBAL DEFAULT   11 QDLDL_solve
     9: 0000000000001120   278 FUNC    GLOBAL DEFAULT   11 QDLDL_etree
```

I am marking as a draft since I still want to do some testing on platforms other than Linux.

### CODE REVIEW

File: include/qdldl.h [hunk 1]
```diff
@@ -5,6 +5,14 @@
 #include "qdldl_types.h"
 #include "qdldl_version.h"
 
+// Define the function attributes that are needed to mark functions as being
+// visible for linking in the shared library version of QDLDL
+#if defined(_WIN32) || defined(_WIN64)
```
Comment (traversaro) [inline include/qdldl.h:10]: As far as I know, on all the platforms in which `_WIN64` is defined in all the platforms `_WIN32` is also defined (see https://docs.microsoft.com/en-us/cpp/preprocessor/predefined-macros?view=msvc-160), so probably it should be sufficient to leave `defined(_WIN32)` .

File: include/qdldl.h [hunk 2]
```diff
@@ -5,6 +5,14 @@
...
+#define QDLDL_API_EXPORT __declspec(dllexport)
```
Comment (traversaro) [inline include/qdldl.h:11]: The classic use of the Windows visibility attribute for shared libraries is that `QDLDL_API_EXPORT ` is defined as `__declspec(dllexport)` only when the header code is compiled as part of the `qdldl` compilation unit, but then when the header code is included in the compilation unit of the downstream library, `QDLDL_API_EXPORT` should be defined as `__declspec(dllimport)` . Furthermore, when the library is compiled as static `QDLDL_API_EXPORT` should always been defined as empty. 

If you just want to use a single header, the boilerplate in the Step-by-Step section of https://gcc.gnu.org/wiki/Visibility may be useful, while if you are ok with generating the header automatically, you can use CMake's [GenerateExportHeader](https://cmake.org/cmake/help/latest/module/GenerateExportHeader.html) module.

File: include/qdldl.h (hunk 1 above)
Comment (imciner2) [reply include/qdldl.h]: I was copying the way Julia had written this define (they used both those). I have cleaned it up.

File: include/qdldl.h (hunk 2 above)
Comment (imciner2) [reply include/qdldl.h]: Technically the use of `__declspec(dllimport)` is not required (https://docs.microsoft.com/en-us/cpp/build/importing-into-an-application-using-declspec-dllimport?view=msvc-160 says "Using `__declspec(dllimport)` is optional on function declarations, but the compiler produces more efficient code if you use this keyword."), but I have updated this to switch between exporting and importing when needed.

**Review by bstellato — APPROVED** [review]

**Pull request merged**